File: pegasusX/apps/warehouse-portal/lib/forecast-confidence.ts
import type { ForecastConfidence } from '@pegasusx/types';

function asNumber(value: unknown): number | undefined {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string' && value.trim() !== '') {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return undefined;
}

function asString(value: unknown): string | undefined {
  if (typeof value === 'string' && value.trim() !== '') return value;
  return undefined;
}

/** Parse planning-brain confidence fields from demand_breakdown JSON. */
export function parseForecastConfidence(
  breakdown?: Record<string, unknown> | null,
): ForecastConfidence | null {
  if (!breakdown || typeof breakdown !== 'object') return null;

  const low = asNumber(breakdown.low_units);
  const high = asNumber(breakdown.high_units);
  let confidencePct = asNumber(breakdown.confidence_pct);
  const rawConfidence = breakdown.confidence;
  if (confidencePct == null && typeof rawConfidence === 'number') {
    confidencePct = rawConfidence <= 1
      ? Math.round(rawConfidence * 100)
      : Math.round(rawConfidence);
  }
  const baselineSource = asString(breakdown.baseline_source);
  const blockedReason = asString(breakdown.blocked_reason);
  const label = asString(breakdown.label) as ForecastConfidence['label'];

  const predicted = asNumber(breakdown.predictedQty) ?? asNumber(breakdown.predicted_qty);
  const derivedLow = low ?? (predicted != null ? Math.max(0, Math.floor(predicted * 0.9)) : undefined);
  const derivedHigh = high ?? (predicted != null ? Math.ceil(predicted * 1.1) : undefined);

  if (
    derivedLow == null
    && derivedHigh == null
    && confidencePct == null
    && !blockedReason
    && !baselineSource
    && !label
  ) {
    return null;
  }

  return {
    low_units: derivedLow,
    high_units: derivedHigh ?? derivedLow,
    confidence_pct: confidencePct,
    baseline_source: baselineSource,
    blocked_reason: blockedReason,
    label,
  };
}

export function isForecastBlocked(confidence?: ForecastConfidence | null): boolean {
  if (!confidence) return false;
  return Boolean(confidence.blocked_reason) || confidence.label === 'insufficient_history';
}

export function isSeasonalTemplateActive(confidence?: ForecastConfidence | null): boolean {
  return confidence?.baseline_source === 'seasonal_template';
}

export function formatSourceBadge(source?: string): string {
  if (!source) return '';
  switch (source) {
    case 'ml': return 'ML';
    case 'moving_average': return 'Baseline';
    case 'seasonal_template': return 'Seasonal';
    case 'mixed': return 'Mixed';
    default: return source.replaceAll('_', ' ');
  }
}

File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/util/ForecastConfidence.kt
package com.pegasusx.warehouse.util

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull

data class ForecastConfidenceData(
    val lowUnits: Long? = null,
    val highUnits: Long? = null,
    val confidencePct: Int? = null,
    val baselineSource: String? = null,
    val blockedReason: String? = null,
    val label: String? = null,
) {
    val blocked: Boolean
        get() = !blockedReason.isNullOrBlank() || label == "insufficient_history"

    val seasonalActive: Boolean
        get() = baselineSource == "seasonal_template"
}

fun parseForecastConfidence(breakdown: JsonObject?): ForecastConfidenceData? {
    if (breakdown == null || breakdown.isEmpty()) return null
    val low = jsonLong(breakdown["low_units"])
    val high = jsonLong(breakdown["high_units"])
    var confidencePct = jsonLong(breakdown["confidence_pct"])?.toInt()
    if (confidencePct == null) {
        val raw = jsonDouble(breakdown["confidence"])
        if (raw != null) {
            confidencePct = if (raw <= 1.0) (raw * 100).toInt() else raw.toInt()
        }
    }
    val baselineSource = jsonString(breakdown["baseline_source"])
    val blockedReason = jsonString(breakdown["blocked_reason"])
    val label = jsonString(breakdown["label"])
    val predicted = jsonLong(breakdown["predictedQty"]) ?: jsonLong(breakdown["predicted_qty"])
    val derivedLow = low ?: predicted?.let { (it * 0.9).toLong().coerceAtLeast(0) }
    val derivedHigh = high ?: predicted?.let { (it * 1.1).toLong().coerceAtLeast(it) }

    if (derivedLow == null && derivedHigh == null && confidencePct == null
        && blockedReason.isNullOrBlank() && baselineSource.isNullOrBlank() && label.isNullOrBlank()
    ) {
        return null
    }
    return ForecastConfidenceData(
        lowUnits = derivedLow,
        highUnits = derivedHigh ?: derivedLow,
        confidencePct = confidencePct,
        baselineSource = baselineSource,
        blockedReason = blockedReason,
        label = label,
    )
}

fun formatSourceBadge(source: String?): String = when (source) {
    "ml" -> "ML"
    "moving_average" -> "Baseline"
    "seasonal_template" -> "Seasonal"
    "mixed" -> "Mixed"
    null, "" -> ""
    else -> source.replace('_', ' ')
}

private fun jsonLong(element: JsonElement?): Long? {
    val primitive = element as? JsonPrimitive ?: return null
    return primitive.longOrNull ?: primitive.content.trim('"').toLongOrNull()
}

private fun jsonDouble(element: JsonElement?): Double? {
    val primitive = element as? JsonPrimitive ?: return null
    return primitive.doubleOrNull ?: primitive.content.trim('"').toDoubleOrNull()
}

private fun jsonString(element: JsonElement?): String? {
    val primitive = element as? JsonPrimitive ?: return null
    val raw = primitive.content.trim('"')
    return raw.ifBlank { null }
}
